diff -up bind-9.7.2-P3/bin/dig/dig.1.rh640538 bind-9.7.2-P3/bin/dig/dig.1
diff -up bind-9.7.2-P3/bin/dig/dig.docbook.rh640538 bind-9.7.2-P3/bin/dig/dig.docbook
--- bind-9.7.2-P3/bin/dig/dig.docbook.rh640538	2011-01-06 12:22:03.259384206 +0100
+++ bind-9.7.2-P3/bin/dig/dig.docbook	2011-01-06 12:26:25.553903983 +0100
@@ -920,6 +920,40 @@ dig +qr www.isc.org any -x 127.0.0.1 isc
   </refsect1>
 
   <refsect1>
+    <title>RETURN CODES</title>
+    <para>
+      <command>Dig</command> return codes are:
+      <variablelist>
+        <varlistentry>
+          <listitem>
+            <para>0: Everything went well, including things like NXDOMAIN</para>
+          </listitem>
+        </varlistentry>
+        <varlistentry>
+          <listitem>
+            <para>1: Usage error</para>
+          </listitem>
+        </varlistentry>
+        <varlistentry>
+          <listitem>
+            <para>8: Couldn't open batch file</para>
+          </listitem>
+        </varlistentry>
+        <varlistentry>
+          <listitem>
+            <para>9: No reply from server</para>
+          </listitem>
+        </varlistentry>
+        <varlistentry>
+          <listitem>
+            <para>10: Internal error</para>
+          </listitem>
+        </varlistentry>
+      </variablelist>
+    </para>
+  </refsect1>
+
+  <refsect1>
     <title>FILES</title>
     <para><filename>/etc/resolv.conf</filename>
     </para>
