diff -up bind-9.8.2rc1/lib/dns/rrl.c.slip-1-fix bind-9.8.2rc1/lib/dns/rrl.c
--- bind-9.8.2rc1/lib/dns/rrl.c.slip-1-fix	2014-06-23 17:46:55.473827749 +0200
+++ bind-9.8.2rc1/lib/dns/rrl.c	2014-06-23 17:48:31.655691394 +0200
@@ -747,6 +747,8 @@ debit_rrl_entry(dns_rrl_t *rrl, dns_rrl_
 	}
 	if (slip != 0 && e->key.s.rtype != DNS_RRL_RTYPE_ALL) {
 		if (e->slip_cnt++ == 0) {
+			if ((int) e->slip_cnt >= slip)
+				e->slip_cnt = 0;
 			if (isc_log_wouldlog(dns_lctx, DNS_RRL_LOG_DEBUG3))
 				debit_log(e, age, "slip");
 			return (DNS_RRL_RESULT_SLIP);
