diff -up bind-9.8.2rc1/bin/dig/dighost.c.rh734458 bind-9.8.2rc1/bin/dig/dighost.c
--- bind-9.8.2rc1/bin/dig/dighost.c.rh734458	2012-02-15 15:09:59.537733633 +0100
+++ bind-9.8.2rc1/bin/dig/dighost.c	2012-02-15 15:12:18.323390093 +0100
@@ -2717,8 +2717,13 @@ connect_timeout(isc_task_t *task, isc_ev
 		}
 	} else {
 		fputs(l->cmdline, stdout);
-		printf(";; connection timed out; no servers could be "
-		       "reached\n");
+		if (!next_origin(query)) {
+			printf(";; connection timed out; no servers could be "
+			       "reached\n");
+		} else {
+			printf(";; connection timed out; trying next "
+			       "origin\n");
+		}
 		cancel_lookup(l);
 		check_next_lookup(l);
 		if (exitcode < 9)
@@ -3364,7 +3369,8 @@ recv_done(isc_task_t *task, isc_event_t
 		return;
 	}
 	if ((msg->rcode == dns_rcode_servfail && !l->servfail_stops) ||
-	    (check_ra && (msg->flags & DNS_MESSAGEFLAG_RA) == 0 && l->recurse))
+	    (check_ra && (msg->flags & DNS_MESSAGEFLAG_RA) == 0 &&
+	     msg->rcode != dns_rcode_noerror && l->recurse))
 	{
 		dig_query_t *next = ISC_LIST_NEXT(query, link);
 		if (l->current_query == query)
