diff -up bind-9.8.1rc1/bin/tests/system/dlzexternal/Makefile.in.dlz_buildfix bind-9.8.1rc1/bin/tests/system/dlzexternal/Makefile.in
diff -up bind-9.8.1rc1/bin/tests/system/Makefile.in.dlz_buildfix bind-9.8.1rc1/bin/tests/system/Makefile.in
--- bind-9.8.1rc1/bin/tests/system/Makefile.in.dlz_buildfix	2011-08-31 17:04:34.466423526 +0200
+++ bind-9.8.1rc1/bin/tests/system/Makefile.in	2011-08-31 17:04:38.512423247 +0200
@@ -21,7 +21,7 @@ top_srcdir =	@top_srcdir@
 
 @BIND9_MAKE_INCLUDES@
 
-SUBDIRS =	dlzexternal filter-aaaa lwresd rpz tkey tsiggss
+SUBDIRS =	filter-aaaa lwresd rpz tkey tsiggss
 TARGETS =
 
 @BIND9_MAKE_RULES@
 
